runtime/src/chpl-topo.c

```diff
@@ -148,12 +148,30 @@ void chpl_topo_init(void) {
   }
 
   //
-  // Note: This will find not only traditional CPU+memory NUMA nodes
-  //       but also memory-only nodes, so for example on KNL it may
-  //       not match chpl_task_getNumSublocales().  We will need to
-  //       revisit this in the future.
+  // Find the NUMA nodes, that is, the objects at numaLevel that also
+  // have CPUs.  This is as opposed to things like Xeon Phi HBM, which
+  // is memory-only, no CPUs.
   //
-  numNumaDomains = hwloc_get_nbobjs_by_depth(topology, numaLevel);
+  {
+    hwloc_cpuset_t cpuset;
+    hwloc_obj_t obj;
+
+    if ((cpuset = hwloc_bitmap_alloc()) == NULL) {
+      report_error("hwloc_bitmap_alloc()", errno);
+    }
+
+    numNumaDomains = 0;
+    for (obj = hwloc_get_next_obj_by_depth(topology, numaLevel, NULL);
+         obj != NULL;
+         obj = hwloc_get_next_obj_by_depth(topology, numaLevel, obj)) {
+      hwloc_bitmap_and(cpuset, obj->online_cpuset, obj->allowed_cpuset);
+      if (!hwloc_bitmap_iszero(cpuset)) {
+        numNumaDomains++;
+      }
+    }
+
+    hwloc_bitmap_free(cpuset);
+  }
```

Comment:
The latest hwloc release that we picked up has a working hwloc_get_nbobjs_inside_cpuset_by_depth that could be used instead of this code.  In that past, that function was buggy, but the release we have now fixes it.